src/ml_board/frontend/Dashboard/src/components/modelCard/PipelineCard/utils.ts

```diff
@@ -0,0 +1,63 @@
+import { Edge, MarkerType, Node } from "reactflow";
+import { IPipelineNode, IReactFlowPipeline } from "./interface";
+import Dagre, { Label } from '@dagrejs/dagre';
+
+const g = new Dagre.graphlib.Graph().setDefaultEdgeLabel(() => ({}));
+
+export const getLayoutedNodes = (nodes: Node[], edges: Edge[], options: { direction: string } = { direction: "TB" }) => {
```

Comment:
I found the layout issue -- why nodes are not laid out properly::

You need to make changes to this function. Here is an example snippet. Please play round to find out the correct positions of x & y.

`

export const getLayoutedNodes = (nodes: Node[], edges: Edge[], options: { direction: string } = { direction: "TB" }) => {
    if (!(nodes.length > 0 && edges.length > 0))
        return { nodes, edges }; // return imidiately if any is emtpy

    g.setGraph({ rankdir: options.direction });
    edges.forEach((edge) => g.setEdge(edge.source, edge.target));
    nodes.forEach((node) => g.setNode(node.id, node as Label));
    Dagre.layout(g);
    let i=0
    return {
        nodes: nodes.map((node: any) => {
            const { x, y } = g.node(node.id);
            // Find nodes with the same rank
            const nodesWithSameRank = nodes.filter((n: any) => n.rank === node.rank);
            // If there are nodes with the same rank (excluding the current node), update x-coordinate
            if (nodesWithSameRank.length > 1) {
                const sameRankNode: any = nodesWithSameRank.find((n) => n.id !== node.id);
                if (sameRankNode) {
                    node.x = sameRankNode.x; 
                    node.position.x = sameRankNode.position.x
                    // ------ PLAY AROUND TO FIND RIGHT CONFIGURATION ------ 
                    i=i+1
                    let node_with_new_corodinates = { ...node, x:x+(100*i), y: y+(100*i), position: { x:x+(100*i), y: y+(100*i) } }
                    return node_with_new_corodinates;
                }
                else {
                    // ------ PLAY AROUND TO FIND RIGHT CONFIGURATION ------ 
                    i=i+1
                    let node_with_new_corodinates = { ...node, x:x+(100*i), y: y+(100*i), position: { x:x+(100*i), y: y+(100*i) } }
                    return node_with_new_corodinates;
                }
            }
            else {
                // ------ PLAY AROUND TO FIND RIGHT CONFIGURATION ------ 
                i=i+1
                let node_with_new_corodinates = { ...node, x:x+(100*i), y: y+(100*i), position: { x:x+(100*i), y: y+(100*i) } }
                return node_with_new_corodinates;
            }
        }),
        edges: edges
    };
};

`

Reply:
optimization aside this code is not solving the problem for me, should I change something else somewhere ??
and why `100`? do you assume that a node width and height is 100px?